File: 3. Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
         int currentLength = 0;
        int maxLength = 0;
        int presence[128] = {0};
        
        for (int i = 0; i < s.length(); i++) {
            currentLength = min(i + 1 - presence[s[i]], currentLength + 1);
            maxLength = max(currentLength, maxLength);
            presence[s[i]] = i + 1;
        }

        return maxLength;
    }
};

File: Leetcode/3. Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:  
            return 0

        p1 = 0
        freq = {}
        n = len(s)
        res = 0

        for p2 in range(n):
            if s[p2] in freq and freq[s[p2]] > 0:
                while s[p1] != s[p2]:
                    freq[s[p1]] -= 1
                    p1 += 1
                p1 += 1  
            else:
                freq[s[p2]] = 1
                res = max(res,p2 - p1 + 1)

        return res
